gtest/src/manager.rs

```diff
@@ -30,17 +30,82 @@ use gear_core::{
 use std::collections::{BTreeMap, VecDeque};
 use std::time::{SystemTime, UNIX_EPOCH};
 
-#[derive(Clone, Debug)]
-pub(crate) enum ProgramState {
-    Initialized,
-    Uninitialized(Option<MessageId>),
-    FailedInitialization,
+#[derive(Debug)]
+pub(crate) enum Actor {
+    Initialized(Program),
+    // Contract: program is always `Some`, option is used to take ownership
+    Uninitialized(Option<MessageId>, Option<Program>),
+    Dormant,
+}
+
+impl Actor {
+    pub(crate) fn new(prog: Program) -> Self {
+        Actor::Uninitialized(None, Some(prog))
+    }
+
+    // #Panics
+    // If actor is initialized or dormant
+    fn transform_to_initialized(&mut self) {
+        assert!(
+            self.is_uninitialized(),
+            "can't transmute actor, which isn't uninitialized"
+        );
+        if let Actor::Uninitialized(_, prog) = self {
+            let prog = prog
+                .take()
+                .expect("actor storage contains only `Some` values by contract");
+            *self = Actor::Initialized(prog);
+        }
+    }
+
+    fn is_dormant(&self) -> bool {
+        matches!(self, Actor::Dormant)
+    }
+
+    fn is_uninitialized(&self) -> bool {
+        matches!(self, Actor::Uninitialized(..))
+    }
+
+    fn as_mut_core_prog(&mut self) -> Option<&mut CoreProgram> {
+        match self {
+            Actor::Initialized(Program::Genuine(prog)) => Some(prog),
+            _ => None,
+        }
+    }
+
+    fn pull_mock(&mut self) -> Option<Box<dyn WasmProgram>> {
+        match self {
+            Actor::Initialized(Program::Mock(mock)) => mock.take(),
+            Actor::Uninitialized(_, Some(Program::Mock(mock))) => mock.take(),
+            _ => None,
+        }
+    }
+
+    fn pull_executable_actor(&mut self, balance: u128) -> Option<ExecutableActor> {
+        let program = match self {
+            Actor::Initialized(Program::Genuine(program)) => Some(program.clone()),
+            Actor::Uninitialized(_, Some(Program::Genuine(program))) => Some(program.clone()),
+            _ => None,
+        };
+        program.map(|program| ExecutableActor { program, balance })
+    }
 }
 
 #[derive(Debug)]
 pub(crate) enum Program {
-    Core(CoreProgram),
-    Mock(Box<dyn WasmProgram>),
+    Genuine(CoreProgram),
+    // Contract: is always `Some`, option is used to take ownership
+    Mock(Option<Box<dyn WasmProgram>>),
+}
+
+impl Program {
+    pub(crate) fn new_genuine(prog: CoreProgram) -> Self {
```

Comment:
I don't like the word `genuine` here, it may confuse.
```suggestion
    pub(crate) fn new(prog: CoreProgram) -> Self {
```

Reply:
done